### ISSUE TITLE
Why is ID -> fresh defaulting specified for function export abbreviation?

### ISSUE THREAD

**Issue by codefromthecrypt** [WebAssembly/spec#1417]
Post MVP, text clarified ID handling on abbreviated/inlined function export, that an ID is to be generated if not present
![Screen Shot 2022-02-07 at 8 35 03 AM](https://user-images.githubusercontent.com/64215/152708661-29511b04-2095-4e30-950f-765deb07f594.png)

This is confusing as it seems an implementation detail. IDs in exports don't serve any purpose except a substitute for a numeric position in the function index namespace. IOTW, one can write the expansion rules and keep position numerically in the index namespace regardless of any ID. Also, I can't see any implementation manifesting different outcomes (ex in the function name section) regardless of whether a function with an inlined export had an ID and regardless of if there were multiple inlined exports or not.

Concretely, nothing ends up in the name section in either the repeated export or one that repeated and never had an ID. (for latest wabt or wasm-tools)
```
(module
  (func $foo (export "foo") (export "bar"))
  (func (export "baz") (export "qux"))
)
```

TL;DR; Why is this specified and what are implementations supposed to do differently knowing it? If there isn't any impact and it is just for implementation advise, could this be reworded, or even removed?

**Comment by codefromthecrypt** [comment]
In looking closely at this, my guess is that this ID defaulting is an implementation hint on how to deal with the out-of-order problem caused by module functions becoming imports later due to abbreviation. One elegant part of the spec was that imports are before module-defined functions. At a glance, this could allow some single-pass parsing to work well... that is, if it weren't for the odd (IMHO) abbreviation where `(module (func (import` is an allowable subsitute for `(module (import (func`. Basically, it is this abbreviation which negates the parsing value of imports first, as any func later can be a "sneaky import". Effectively you have to assume in parsing that the end of the file has one of these oddballs, forcing you to code as if the imports first rule wasn't there in the first place (at least from a single-pass parsing POV).

Commentary on what I feel was an unwise addition to the spec, the point remains that this impl advice is equally odd without qualification as to why it exists. It is one way an impl can deal with the "sneaky import" problem, but unless you know that it is a problem, it feel off. I might recommend documenting how the spec by allowing sneaky imports, requires approaches such as this to keep the index coherent.

**Comment by rossberg** [comment]
> Why is this specified and what are implementations supposed to do differently knowing it? If there isn't any impact and it is just for implementation advise, could this be reworded, or even removed?

The expansion has to refer to the right definition. Using a fresh name is a simple, declarative way to specify the outcome without requiring any form of global knowledge in the rule (except for the benign freshness property).

The only alternative would be to specify a precise ordering on how expansions of abbreviations are performed and somehow track the (current) size of the full index space through it, which is tricky, since expansions is defined before we even determine index spaces. (An algorithmic implementation will need two passes to do this.)

> Concretely, nothing ends up in the name section

The name section is part of the binary format and unrelated to the text grammar. The spec does not determine or prescribe what to put in a name section.

> Basically, it is this abbreviation which negates the parsing value of imports first, as any func later can be a "sneaky import".

Ah, no, the abbreviated import syntax is still subject to the ordering constraint [imposed on "module composition"](https://webassembly.github.io/spec/core/text/modules.html#text-module).

In general, abbreviations are well-formed if and only if their expansions are.

**Comment by codefromthecrypt** [comment]
>Ah, no, the abbreviated import syntax is still subject to the ordering constraint [imposed on "module composition"](https://webassembly.github.io/spec/core/text/modules.html#text-module).
> In general, abbreviations are well-formed if and only if their expansions are.

hmm so you mean we should fail parsing if what I refer to as a "sneaky import" ex. `(func (import` is declared after a normal `(func`?

**Comment by rossberg** [comment]
Yes. I think there are tests for that, too.
```
> (module (func) (func (import "" "")))
stdin:1.16-1.37: syntax error: import after function definition
```

**Comment by codefromthecrypt** [comment]
ok thanks. You've answered my question about why on the IDs and also clarified that the abbreviation isn't a loophole to do "sneaky imports" appreciate your time on this!
